gdeep/trainer/trainer.py

```diff
@@ -77,6 +89,149 @@ def wrapper(*args, **kwargs):
+    _DP = auto()
+    FSDP = auto()
+    PIPELINE = auto()
+
+    def from_str(string):
+            if string.upper() == "FSDP":
+                return ParallelismType.FSDP
+            elif string.upper() == "PIPELINE":
+                return ParallelismType.PIPELINE
+            else:
+                return ParallelismType._NONE
+
+class Parallelism:
+    """Stores the necessary informations to perform parallel training using the Trainer class. Some attributes are only used for FSDP (p_type = DDP, FSDP_ZERO2, FSDP_ZERO3) or pipeline_tool (p_type = PIPELINE) and are indicated as such in their description. No indication means they are used by both tools
+    Args:
+        p_type:
+            Type of parallelism training algorithm to use
+        devices:
+            Indices of the available GPUs to use for training
+        nb_device:
+            Actual number of GPUs to use from the ones referenced in devices. Asking for more GPUs than referenced in devices will result in an exception
+        config_fsdp:
+            [FSDP] FSDP configuration dictionary. See pytorch's FSDP documentation for more details
+        config_mha:
+            [pipeline_tool] Multihead Attention layers configuration in the transformer (if the model used is a transformer)
+        pipeline_chunks:
+            [pipeline_tool] Number of chunks to split the model into for pipelining
+    """
+    def __init__(self,
+                p_type: ParallelismType,
+                devices: Tuple[int] = None,
+                nb_device: int = 0,
+                config_fsdp: Dict[str, Any] = {},
+                config_mha: List[Dict[str, Any]] = [],
+                pipeline_chunks: int = 4) -> None:
+        self.p_type = p_type
+        self.world_size = nb_device
+        self.rank = 0
+        self.config_mha = config_mha
+        self.pipeline_chunks = pipeline_chunks
+        self.config_fsdp = config_fsdp
+
+        if p_type == ParallelismType.FSDP and \
+            "sharding_strategy" in config_fsdp.keys() and \
+            config_fsdp["sharding_strategy"] == ShardingStrategy.NO_SHARD and \
+            not "auto_wrap_policy" in config_fsdp.keys():
+            print("WARNING: You seem to be trying to use FSDP with sharding without providing any wrapping policy. \
+                  This will result in the model being unable to be sharded between the GPUs. Only use FSDP without \
+                  wrapping policy with sharding_strategy == ShardingStrategy.NO_SHARD")
+
+        if devices is None:
+            devices = list(range(torch.cuda.device_count()))
+        self.devices = [torch.device('cuda', x) for x in devices] # Convert device indices into torch devices
+
+        # Verify proper use of devices
+        if self.world_size > len(self.devices):
+            raise ValueError("Cannot use more devices than those referenced in devices")
+        if self.world_size < 1:
+            self.world_size = len(self.devices)
+
+
+def setup_env():
```

Comment:
Why do we need to set env variables here? This looks very error-prone.

Reply:
This is necessary for parallel training with RPC.
There is no choice but defining env variables.